core/node/node_framework/src/service/context_traits.rs

```diff
@@ -0,0 +1,133 @@
+use crate::{resource::Resource, service::context::ServiceContext, wiring_layer::WiringError};
+
+/// Trait used as input for wiring layers, aiming to provide all the resources the layer needs for wiring.
+///
+/// For most cases, the most conevenient way to implement this trait is to use the `#[derive(FromContext)]`.
+/// Otherwise, the trait has several blanket implementations (including the implementation for `()` and `Option`).
+///
+/// # Example
+///
+/// ```
+/// use zksync_node_framework::FromContext;
+/// # #[derive(Clone)]
+/// # struct MandatoryResource;
+/// # impl zksync_node_framework::resource::Resource for MandatoryResource { fn name() -> String { "a".into() } }
+/// # #[derive(Clone)]
+/// # struct OptionalResource;
+/// # impl zksync_node_framework::resource::Resource for OptionalResource { fn name() -> String { "b".into() } }
+/// # #[derive(Default, Clone)]
+/// # struct ResourceWithDefault;
+/// # impl zksync_node_framework::resource::Resource for ResourceWithDefault { fn name() -> String { "c".into() } }
+/// #[derive(FromContext)]
+/// struct MyWiringLayerInput {
+///     // The following field _must_ be present in the context.
+///     mandatory_resource: MandatoryResource,
+///     // The following field is optional.
+///     // If will be `None` if there is no such resource in the context.
+///     optional_resource: Option<OptionalResource>,
+///     // The following field is guaranteed to fetch the value from the context.
+///     // In case the value is missing, a default value will be added to the context.
+///     #[context(default)]
+///     resource_with_default: ResourceWithDefault,
+/// }
+/// ```
+pub trait FromContext: Sized {
+    fn from_context(context: &mut ServiceContext<'_>) -> Result<Self, WiringError>;
+}
+
+impl<T: Resource + Clone> FromContext for T {
+    fn from_context(context: &mut ServiceContext<'_>) -> Result<Self, WiringError> {
+        context.get_resource::<T>()
+    }
+}
+
+impl FromContext for () {
+    fn from_context(_context: &mut ServiceContext<'_>) -> Result<Self, WiringError> {
+        Ok(())
+    }
+}
+
+impl<T: Resource + Clone> FromContext for Option<T> {
```

Comment:
Bikeshedding: Wouldn't it be possible to generalize this implementation to `T: FromContext` (ditto for `impl IntoContext for Option<T>` below)? It would make these traits more composable.